core/src/main/java/tech/tablesaw/io/ReadOptions.java

```diff
@@ -327,8 +358,133 @@ public Builder minimizeColumnSizes() {
       return this;
     }
 
+    /**
+     * Provide column types for all columns skipping autodetect column type logic. The array must
+     * contain a ColumnType for each column in the table. An error will be thrown if they don't
+     * match up
+     */
+    public Builder columnTypes(ColumnType[] columnTypes) {
+      if (columnTypeOptionsAlreadySet()) {
+        throw new IllegalStateException("columnTypes already set");
+      }
+      this.columnTypes = columnTypes;
+      return this;
+    }
+
+    /**
+     * Provide a function that determines ColumnType for all column names. To provide only for some
+     * use {@link #columnTypesPartial(Function)}
+     *
+     * <p>This method is generally more efficient because it skips column type detection
+     */
+    public Builder columnTypes(Function<String, ColumnType> columnTypeFunction) {
+      if (columnTypeOptionsAlreadySet()) {
+        throw new IllegalStateException("columnTypes already set");
+      }
+      this.completeColumnTypeFunction = columnTypeFunction;
+      return this;
+    }
+
+    /**
+     * Provide a function that determines ColumnType for some column names. To provide for all
+     * column names use {@link #columnTypes(Function)} that generally is more efficient because it
+     * skips column type detection
+     */
+    public Builder columnTypesPartial(Function<String, Optional<ColumnType>> columnTypeFunction) {
```

Comment:
Do we need both `columnTypes` and `columnTypesPartial`? Could we instead have the reader check if all column names were specified then it will skip detection?

Reply:
@benmccann It's what I commented to @lwhite1 in comment 6 days ago.
Technically it's possible but it will require a refactor in the way we buffer and process files and we agreed to maintain that interface differentiating the complete cases and the partial ones.
I think that maintaining the partial is not a bad idea because it has important implications in performance for big files, and making the user aware can be a good thing.

In case we want to suppress the partials, I think that the best will be that in this PR always run type detection logic for all the new cases.  And open an issue to refactor the caching mechanism to allow a more intelligent approach for the new cases when all the column types can be determined.